src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala

```diff
@@ -246,9 +245,42 @@ trait SubmissionComponent {
     }
 
     def loadSubmissionWorkflows(submissionId: UUID): ReadAction[Seq[Workflow]] = {
-      workflowQuery.findWorkflowsBySubmissionId(submissionId).result.flatMap(recs => DBIO.sequence(recs.map { rec =>
-        workflowQuery.get(rec.id) map(wf => wf.get)
-      }))
+      // get all workflows for this submission, with their entity and messages:
+      // WORKFLOW left join ENTITY left outer join WORKFLOW_MESSAGE
+      val workflows = for {
+        ((workflow, entity), message) <-
+          workflowQuery.filter(_.submissionId===submissionId) join
+            entityQuery on (_.workflowEntityId === _.id) joinLeft
+            workflowMessageQuery on (_._1.id === _.workflowId)
+      } yield (workflow, entity, message)
+
+      workflows.result.map{recs =>
+        // group the workflow/entity/message records by workflowId, which creates a map keyed by workflowId
+        val recordsByWorkflowId: Map[Long, Seq[(WorkflowRecord, EntityRecord, Option[WorkflowMessageRecord])]] = recs.groupBy(_._1.id)
```

Comment:
it is more readable to use a case statement instead of tupleese

 recs.groupBy { case (workflow, entity, message) => workflow.id }

then the comment above is not as necessary